src/mono/mono/mini/aot-compiler.c

```diff
@@ -14355,6 +14392,195 @@ add_interp_in_wrapper_for_sig (MonoAotCompile *acfg, MonoMethodSignature *sig)
+		parsed = FALSE;
+		goto early_exit;
+	}
+
+	*module_name_ptr = g_strdup (direct_pinvoke_split[0]);
+	if (!*module_name_ptr) {
+		aot_printerrf (acfg, "Failed to strdup the module name portion of direct pinvoke '%s'.\n", direct_pinvoke);
+		parsed = FALSE;
+	}
+
+	// ENTRYPOINT can be NULL if Direct PInvoke is just MODULE
+	*entrypoint_name_ptr = g_strdup (direct_pinvoke_split[1]);
+	if (direct_pinvoke_split[1] && !*entrypoint_name_ptr) {
+		aot_printerrf (acfg, "Failed to strdup the entrypoint name portion of direct pinvoke '%s'.\n", direct_pinvoke);
+		parsed = FALSE;
+	}
+
+	g_strfreev (direct_pinvoke_split);
+	direct_pinvoke_split = NULL;
+
+early_exit:
+	return parsed;
+}
+
+//---------------------------------------------------------------------------------------
+//
+// add_direct_pinvoke adds the module and entrypoint of a specified direct pinvoke to the
+// MonoAotCompile instance's HashTable of module/entrypoint entries.
+// It is presumed that module_name_ptr and entrypoint_name_ptr point to valid strings.
+// It transfers ownership of the module_name and entrypoint_name strings data to the HashTable
+// This function takes ownership of the module and entrypoint strings pointers regardless of
+// success or failure (in which it frees the memory) and sets them to NULL.
+//
+// Arguments:
+//  * acfg - the MonoAotCompiler instance
+//  ** module_name_ptr - the pointer to the module name (assumed not NULL)
+//  ** entrypoint_name_ptr - the pointer to the entrypoint name
+//
+// Return Value:
+//  gboolean corresponding to whether or not the direct_pinvoke was successfully added
+//  or if an error occurred.
+//
+
+static gboolean
+add_direct_pinvoke (MonoAotCompile *acfg, char **module_name_ptr, char **entrypoint_name_ptr)
+{
+	gboolean success = TRUE;
+	GHashTable *entrypoints;
+	// If there is an entry for the module
+	if (g_hash_table_lookup_extended (acfg->direct_pinvokes, *module_name_ptr, NULL, (gpointer *)&entrypoints)) {
+		// Not all entrypoints are direct, if specifying a new entrypoint
+		if (entrypoints && *entrypoint_name_ptr && !g_hash_table_contains (entrypoints, *entrypoint_name_ptr))
+			g_hash_table_insert (entrypoints, *entrypoint_name_ptr, NULL);
+	// New entry for module, All entrypoints are direct
+	} else if (!*entrypoint_name_ptr) {
+		g_hash_table_insert (acfg->direct_pinvokes, *module_name_ptr, NULL);
+	// New entry for module, specifying an entrypoint
+	} else {
+		entrypoints = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+		if (!entrypoints) {
+			aot_printerrf (acfg, "Failed to allocate new entrypoints HashTable.\n");
+			g_free (*module_name_ptr);
+			g_free (*entrypoint_name_ptr);
+			success = FALSE;
+		} else {
+			g_hash_table_insert (entrypoints, *entrypoint_name_ptr, NULL);
+			g_hash_table_insert (acfg->direct_pinvokes, *module_name_ptr, entrypoints);
+		}
+	}
+
+	*module_name_ptr = NULL;
+	*entrypoint_name_ptr = NULL;
+	return success;
```

Comment:
This method still leaks *module_name_ptr and *entrypoint_name_ptr in a couple of scenarios:

You need to make sure strings are freed in case where ownership is not transfered, meaning if you don't do a g_hash_table_insert on *module_name_ptr or *entrypoint_name_ptr for the call, then the strings have not been transferred into the hash tables and needs to be freed by this function before returning.

Since the logic in this function is a little messy, you could handle that by setting the pointer to NULL as soon as you have done a g_hash_table_insert, like:

g_hash_table_insert (entrypoints, *entrypoint_name_ptr, NULL);
*entrypoint_name_ptr = NULL

and then before you return you do (note g_free won't do anything if the parameter is NULL):

g_free (*module_name_ptr);
g_free (*entrypoint_name_ptr);

*module_name_ptr = NULL;
*entrypoint_name_ptr = NULL;

Reply:
Ah, thank you for pointing that out. I had mistakenly thought that if the string data was already in the table and that the HashTable already had ownership of the data, but now I'm realizing that theyre not actually the same memory, just that they have the same data, and we still need to free the strings even if we dont insert, as you said. Then I believe the only case where the ownership is not transferred is if there is an entry for the module, and either all entrypoints are direct or the entrypoint being specified is already in the entrypoints HashTable.

Thanks for the tip, I was not sure how to cleanly handle freeing for when we don't insert and not accidentally free when we do insert. Setting the pointers to NULL immediately after transferring ownership makes sense to me to achieve that.